### PR TITLE
Removed static root tree, Getting it from store manager

### PR DESCRIPTION
**Pull request by Rus0**
### Description (*)

Removed static root tree, Getting it from store manager

### Fixed Issues (if relevant)

1. magento/graphql-ce#816: Root category ID

### Manual testing scenarios (*)

1. Get a tree without a category id:
category {
    id
....
}

2. Get a tree in an installation with root_category_id = 2, specifying the number 1, it should return an error

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/graphql-ce?pullRequest=851) <br/>All committers have signed the CLA.

**Comment by lenaorobei** [comment]
We also need to add add `root_category_id` to the `StoreConfig`
``` graphql
type StoreConfig @doc(description: "The type contains information about a store config") {
    ...
    root_category_id: Int @doc(description: "The ID ot root category") @resolver(class: "New\Resolver\Here")
}
```

**Comment by lenaorobei** [comment]
Let's extract the code related to the root category ID to the separate class and reuse it in both resolvers. Example: `\Magento\QuoteGraphQl\Model\Cart\GetCartForUser`.

File: app/code/Magento/CatalogGraphQl/Model/Resolver/CategoryRoot.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+
+    /**
+     * @inheritdoc
+     */
+    public function resolve(Field $field, $context, ResolveInfo $info, array $value = null, array $args = null)
+    {
+
+        return $this->getRootCategoryId();
```
Comment (lenaorobei) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/CategoryRoot.php]: `Magento\CatalogGraphQl\Model\Category\GetRootCategoryId` should be used here.

File: app/code/Magento/CatalogGraphQl/Model/Resolver/CategoryRoot.php [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    /**
+     * @var Products\DataProvider\CategoryTree
+     */
+    private $categoryTree;
+
+    /**
+     * @var ExtractDataFromCategoryTree
+     */
+    private $extractDataFromCategoryTree;
```
Comment (lenaorobei) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/CategoryRoot.php]: Please remove redundant private properties.

File: app/code/Magento/CatalogGraphQl/Model/Category/GetRootCategoryId.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
+<?php
+
```
Comment (lenaorobei) [inline app/code/Magento/CatalogGraphQl/Model/Category/GetRootCategoryId.php]: Please add copyright and and `declare(strict_types=1);`

File: app/code/Magento/CatalogGraphQl/Model/Category/GetRootCategoryId.php [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+namespace Magento\CatalogGraphQl\Model\Category;
+
+class GetRootCategoryId
+{
+
```
Comment (lenaorobei) [inline app/code/Magento/CatalogGraphQl/Model/Category/GetRootCategoryId.php]:
```suggestion
```

File: app/code/Magento/CatalogGraphQl/Model/Category/GetRootCategoryId.php [hunk 5]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+namespace Magento\CatalogGraphQl\Model\Category;
+
```
Comment (lenaorobei) [inline app/code/Magento/CatalogGraphQl/Model/Category/GetRootCategoryId.php]: Please add PHPDoc clock here.

File: app/code/Magento/CatalogGraphQl/Model/Category/GetRootCategoryId.php [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+class GetRootCategoryId
+{
+
+    /**
+     * @var int
+     */
+    private $rootCategoryId = null;
+    /**
+     * @var \Magento\Store\Model\StoreManagerInterface
```
Comment (lenaorobei) [inline app/code/Magento/CatalogGraphQl/Model/Category/GetRootCategoryId.php]: Better to move to the `use` section.

File: app/code/Magento/CatalogGraphQl/Model/Category/GetRootCategoryId.php [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace Magento\CatalogGraphQl\Model\Category;
+
+class GetRootCategoryId
+{
+
+    /**
+     * @var int
+     */
+    private $rootCategoryId = null;
```
Comment (lenaorobei) [inline app/code/Magento/CatalogGraphQl/Model/Category/GetRootCategoryId.php]:
```suggestion
    private $rootCategoryId;
```

File: app/code/Magento/CatalogGraphQl/Model/Category/GetRootCategoryId.php [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+    /**
+     * Get Root Category Id
+     * @return int
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
+     */
+    public function execute()
+    {
+        if ($this->rootCategoryId == null) {
+            $this->rootCategoryId = (int)$this->storeManager->getStore()->getRootCategoryId();
```
Comment (lenaorobei) [inline app/code/Magento/CatalogGraphQl/Model/Category/GetRootCategoryId.php]: Looks like we need to wrap with `try catch`.

**Comment by lenaorobei** [comment]
There are some small code style issues.
```
PHP Code Sniffer detected 2 violation(s): 

FILE: ...code/Magento/CatalogGraphQl/Model/Category/GetRootCategoryId.php
----------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
----------------------------------------------------------------------
 14 | ERROR | [x] Class is missing annotation block
 37 | ERROR | [x] There must be exactly one blank line before tags
----------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------



Failed asserting that 2 matches expected 0.
```

**Comment by Rus0** [comment]
@lenaorobei The failed test is not from anything I did

**Comment by lenaorobei** [comment]
@magento run Functional Tests build

**Comment by Rus0** [comment]
@lenaorobei done! :)

**Comment by TomashKhamlai** [comment]
### Initial testing
| Input value | Expected response | Explanation | Status |
| --- | --- | --- | --- |
| [-2,147,483,648;&nbsp;0] | `Category doesn't exist` | It is not expected</br>to use values from</br> the specified range</br>as a category ID | <span style="color: red">FAIL</span> |
| 1 | `Category doesn't exist` | Expected by PR description | <span style="color: #">PASS</span> |
| [2;&nbsp;2,147,483,647]</br>If exists | `CategoryTree` | Expected by default | <span style="color: green">PASS</span> |
| [2;&nbsp;2,147,483,647]</br>If doesn't exist | `Category doesn't exist` | Expected by default | <span style="color: green">PASS</span> |
| null | `CategoryTree` for Default Category | Expected by PR description | <span style="color: green">PASS</span> |

#### Failure:
**Request**
```graphql
{
  category (id: 0) {
		available_sort_by
    breadcrumbs {
      category_url_key
      category_level
      category_id
      category_name
    }
    children {
      children_count
    }
    children_count
    created_at
    default_sort_by
    description
    display_mode
    filter_price_range
    filter_price_range
    id
    image
    include_in_menu
    is_anchor
    landing_page
    level
    meta_description
    meta_keywords
    meta_title
    name
    path
    path_in_store
    position
    product_count
    products {
      items {
        __typename
      }
      page_info {
        current_page
        page_size
        total_pages
      }
      total_count
    }
    updated_at
    url_key
    url_path
  }
}
```

**Expected:**
```graphql
{
  "errors": [
    {
      "message": "Category doesn't exist",
      "category": "graphql-no-such-entity",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "category"
      ]
    }
  ],
  "data": {
    "category": null
  }
}
```

**Actual:**
```graphql
{
  "data": {
    "category": {
      "available_sort_by": [],
      "breadcrumbs": null,
      "children": [],
      "children_count": "0",
      "created_at": "2019-08-29 09:14:47",
      "default_sort_by": null,
      "description": null,
      "display_mode": "PRODUCTS",
      "filter_price_range": null,
      "id": 2,
      "image": null,
      "include_in_menu": 1,
      "is_anchor": null,
      "landing_page": null,
      "level": 1,
      "meta_description": null,
      "meta_keywords": null,
      "meta_title": null,
      "name": "Default Category",
      "path": "1/2",
      "path_in_store": null,
      "position": 1,
      "product_count": 0,
      "products": {
        "items": [],
        "page_info": {
          "current_page": 1,
          "page_size": 20,
          "total_pages": 0
        },
        "total_count": 0
      },
      "updated_at": "2019-08-29 09:14:47",
      "url_key": null,
      "url_path": null
    }
  }
}
```
Looks like a known bug.

**Comment by Rus0** [comment]
@TomashKhamlai I cannot reproduce the failed test in a clean installation.

Am I missing a step for reproducing?

**Comment by Rus0** [comment]
Oh! Forget it, I did reproduce it, I will ammend the pr

**Comment by TomashKhamlai** [comment]
@lenaorobei, I found an interesting thing :slightly_smiling_face:
After running api-functional tests
`php vendor/phpunit/phpunit/phpunit --configuration dev/tests/api-functional/phpunit_graphql.xml dev/tests/api-functional/testsuite/Magento/GraphQl/Catalog;`
when I send
```graphql
{
  category(id: 2) {
    children_count
  }
}
```
or 
```graphql
{
  category(id: null) {
    children_count
  }
}
```
or
```graphql
{
  category {
    children_count
  }
}
```
I get
```graphql
{
  "data": {
    "category": {
      "children_count": "-32"
    }
  }
}
```
Looks like some fixtures or rollbacks know how to borrow categories. Should I create a ticket about this or it is expected?

**Comment by Rus0** [comment]
I think we need a new ticket, it is a new bug, since it does not match the actual result :thinking:

**Review by lenaorobei — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lenaorobei, thank you for the review.
[ENGCOM-5744](https://jira.corp.magento.com/browse/ENGCOM-5744) has been created to process this Pull Request

**Comment by lenaorobei** [comment]
One bug was discovered on internal performance builds:
 **Steps to reproduce:** 
1. Generate performance fixtures
`php bin/magento setup:performance:generate-fixtures ./setup/performance-toolkit/profiles/ce/medium.xml`
2. Make a query:
```
{
  category(id: 1) {
    name
    id
    level
    description
    path
    path_in_store
    url_key
    url_path
    children {
      id
      description
      default_sort_by
      children {
        id
        description
        level
        children {
          level
          id
          children {
            id
          }
        }
      }
    }
  }
}

```
**Expected result:**
Returned data for default category with children.

**Actual result:**
```
{
  "errors": [
    {
      "message": "Category doesn't exist",
      "category": "graphql-no-such-entity",
      "locations": [
        {
          "line": 1,
          "column": 4
        }
      ],
      "path": [
        "category"
      ]
    }
  ],
  "data": {
    "category": null
  }
}
```

**Review by lenaorobei — CHANGES_REQUESTED** [review]
Please address found bug.

**Review by lenaorobei — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lenaorobei, thank you for the review.
[ENGCOM-5744](https://jira.corp.magento.com/browse/ENGCOM-5744) has been created to process this Pull Request

**Pull request merged**